File: C++/946. Validate Stack Sequences.cpp
class Solution {
public:
    bool validateStackSequences(vector<int>& p, vector<int>& d) {
        int i=0;
        vector<int>a;
        for(int x:p){
            a.push_back(x);
            if(a.back()==d[i]){
                while(i<d.size()){
            if(!a.empty() && a.back()==d[i]){
                i++;
                a.pop_back();
            }
            else
            break;
        }
            }
        }
        while(i<d.size()){
            if(a.back()==d[i]){
                i++;
                a.pop_back();
            }
            else
            return false;
        }
        return a.size()==0;
    }
};

File: JAVA/946. Validate Stack Sequences.java
class Solution {
    public boolean validateStackSequences(int[] p, int[] d) {
        Deque<Integer>a=new ArrayDeque<>();
        int i=0;
        for(int x:p){
            a.push(x);
            if(a.peek()==d[i]){
                while(i<d.length){
                    if(a.size()!=0 && a.peek()==d[i]){
                        i++;
                        a.pop();
                    }
                    else
                    break;
                }
            }
        }
        while(i<d.length){
            if(a.peek()==d[i]){
                i++;
                a.pop();
            }
            else
            return false;
        }
        return a.size()==0;
    }
}